sdks/java/io/couchbase/src/main/java/org/apache/beam/sdk/io/couchbase/CouchbaseIO.java

```diff
@@ -0,0 +1,359 @@
+     * to the target bucket.
+     *
+     * @param options Pipeline options
+     * @return the number of keys to the bucket
+     * @throws Exception throw {@link CouchbaseIOException} when querying Couchbase
+     */
+    @Override
+    public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+      String query = String.format("SELECT RAW COUNT(META().id) FROM `%s`", getBucket().name());
+      LOG.debug(query);
+      N1qlQueryResult result = getBucket().query(N1qlQuery.simple(query));
+      if (!result.finalSuccess()) {
+        throw new CouchbaseIOException(result.errors().get(0).getString("msg"));
+      }
+      itemCount =
+          Integer.parseInt(
+              new String(result.allRows().get(0).byteValue(), Charset.defaultCharset()));
+      return itemCount;
+    }
+
+    @Override
+    public BoundedReader<JsonDocument> createReader(PipelineOptions options) {
+      return new CouchbaseReader(this);
+    }
+
+    void close() {
+      if (cluster != null) {
+        cluster.disconnect();
+      }
+      if (bucket != null) {
+        bucket.close();
+      }
+    }
+  }
+
+  private static class CouchbaseReader extends BoundedSource.BoundedReader<JsonDocument> {
+
+    private final CouchbaseSource source;
+    private Iterator<N1qlQueryRow> keyIterator;
+    private String currentKey;
+
+    CouchbaseReader(CouchbaseSource source) {
+      this.source = source;
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      // Fetch the keys between the range
+      String query =
+          String.format(
+              "SELECT RAW META().id FROM `%s` OFFSET %d LIMIT %d",
+              source.spec.bucket(), source.lowerBound, source.upperBound - source.lowerBound);
+      LOG.debug(
+          String.format(
+              "Couchbase reader [%d, %d): %s", source.lowerBound, source.upperBound, query));
+      N1qlQueryResult result = source.bucket.query(N1qlQuery.simple(query));
+      if (!result.finalSuccess()) {
+        throw new CouchbaseIOException(result.errors().get(0).getString("msg"));
+      }
+      List<N1qlQueryRow> keys = result.allRows();
+      keyIterator = keys.iterator();
+      return advance();
+    }
+
+    @Override
+    public boolean advance() {
+      if (keyIterator.hasNext()) {
+        currentKey = new String(keyIterator.next().byteValue(), Charset.defaultCharset());
+        // Need to remove the replicated quotes around the key (""key"" => "key").
+        // The type of key is limited to String.
+        currentKey = currentKey.substring(1, currentKey.length() - 1);
+        return true;
+      }
+      return false;
+    }
+
+    @Override
+    public JsonDocument getCurrent() throws NoSuchElementException {
+      return source.bucket.get(currentKey);
+    }
+
+    @Override
+    public void close() {
+      // Delegate the disconnection of Couchbase to the method "expand" of CouchbaseSource,
```

Comment:
Better idea to open and deal with the connection lifecycle at the reader level, because the Source does not have a lifecycle.